x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,260 @@
+	// the soax package name with the suffix
+	soaxUsername := soaxPackage + suffix
+	log.Printf("Sending associate command with username: %s\n", soaxUsername)
+
+	streamEndpoint := transport.StreamDialerEndpoint{Dialer: &transport.TCPDialer{}, Address: soaxAddress}
+	client, err := csocks5.NewClient(&streamEndpoint)
+	if err != nil {
+		return err
+	}
+
+	err = client.SetCredentials([]byte(soaxUsername), []byte(soaxPassword))
+	if err != nil {
+		return err
+	}
+
+	client.EnablePacket(&transport.UDPDialer{})
+
+	conn, bindAddr, err := client.ConnectAndRequest(ctx, csocks5.CmdUDPAssociate, "0.0.0.0:0")
+	if err != nil {
+		return err
+	}
+
+	// Start a goroutine to close the connection after a timeout
+	go func() {
+		timeout := viper.GetDuration("udp_timeout")
+		<-time.After(timeout)
+		log.Println("Closing UDP associate connection after timeout")
+		conn.Close()
+	}()
+
+	if err = socks5.SendReply(writer, statute.RepSuccess, convertToNetAddr("udp", bindAddr.IP, bindAddr.Port)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func convertToNetAddr(network string, ip netip.Addr, port uint16) net.Addr {
+	// Convert netip.Addr to net.IP
+	netIP := ip.AsSlice()
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+		// Create a net.TCPAddr
+		tcpAddr := &net.TCPAddr{
+			IP:   netIP,
+			Port: int(port),
+		}
+		return tcpAddr
+	case "udp", "udp4", "udp6":
+		// Create a net.UDPAddr
+		udpAddr := &net.UDPAddr{
+			IP:   netIP,
+			Port: int(port),
+		}
+		return udpAddr
+	}
+	return nil
+}
+
+func loadConfig() error {
+	viper.SetConfigName("config")
+	viper.SetConfigType("yaml")
+	viper.AddConfigPath(".")
+
+	if err := viper.ReadInConfig(); err != nil {
+		return fmt.Errorf("failed to read config file: %w", err)
+	}
+	return nil
+}
+
+func main() {
+
+	if err := loadConfig(); err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
+
+	creds := &StaticCredentialStore{
+		credentials: viper.GetStringMapString("credentials"),
+	}
+
+	// Create the custom authenticator
+	customAuth := CustomAuthenticator{
+		PrefixCredentials: creds,
+	}
+
+	server := socks5.NewServer(
+		socks5.WithAuthMethods([]socks5.Authenticator{customAuth}),
+		socks5.WithLogger(socks5.NewLogger(log.New(os.Stdout, "socks5: ", log.LstdFlags))),
+		socks5.WithDialAndRequest(func(ctx context.Context, network, addr string, req *socks5.Request) (net.Conn, error) {
+			authContext := req.AuthContext
+			if authContext == nil {
+				return nil, errors.New("no auth context available")
+			}
+
+			suffix, ok := authContext.Payload["suffix"]
+			if !ok {
+				return nil, errors.New("no suffix found in auth context")
+			}
+
+			soaxpackage := viper.GetString("soax.package")
+			soaxPassword := viper.GetString("soax.password")
+			soaxAddress := viper.GetString("soax.address")
+
+			// Construct the upstream username by concatenating the base username with the suffix
+			soaxUsername := soaxpackage + suffix // e.g., "package-189365-country-ir-seesionid-..."
+
+			switch network {
```

Comment:
Handle UDP here as well. There's no need to implement an AssociateHandler, since [the library already provides that](https://github.com/things-go/go-socks5/blob/503e7866821a419fe3766751f6d5671390072c14/handle.go#L181).